enhancements/machine-api/control-plane-machineset-vertical-scaling.md

```diff
@@ -0,0 +1,348 @@
+  - `timeWindow`: the time CPU and memory usage have to be below the thresholds,
+    before triggering a scale down event.
+
+A complete configuration in YAML format could then look like this:
+
+```yaml
+controlplaneautoscaling:
+  instanceSizePath: spec.template.spec.providerSpec.value.instanceType
+  instanceSizes:
+  - weight: 1
+    name: r4.large
+  - weight: 2
+    name: r4.xlarge
+  - weight: 4
+    name: r4.2xlarge
+  # [mandatory configuration]
+  syncPeriod: [duration as specified in golang: e.g. 5s, 15s]
+  gracePeriod: [duration as specified in golang: e.g. 5s, 15s]
+  scaledownEnabled: [true|false]
+  scaleUpThresholds:
+    cpuLoadAverage: [0-100 as percentage]
+    memoryUsage: [0-100 as percentage]
+    timeWindow: [duration as specified in golang: e.g. 5m, 15m]
+  scaleDownThresholds:
+    cpuLoadAverage: [0-100 as percentage]
+    memoryUsage: [0-100 as percentage]
+    timeWindow: [duration as specified in golang: e.g. 5m, 15m]
+```
+
+### Data gathering
+
+As specified the operator should not require access to data from operators that
+might not be installed in the cluster (e.g. `prometheus`).
+
+Instead the operator should gather the required data via the [metrics
+API](https://kubernetes.io/docs/tasks/debug/debug-cluster/resource-metrics-pipeline/)
+that is also used for the horizontal and vertical pod autoscalers.
+
+Data available from the metrics API includes:
+
+- CPU usage in `cores`.
+- Memory usage in a SI units.
+- The time window used to aggregate the above mentioned data.
+
+To gather data about the available CPU cores and memory available to control
+plane nodes, the operator should use the kubernetes
+[Node](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#node-v1-core)
+API.
+
+As the window returned by metrics API is likely shorter than required window for
+scaling decisions, the operator should gather its own average to allow making
+the scaling decision based on the metrics data.
+
+### Workflow for automatic scaling
+
+Automatic scaling should be triggered the same as [horizontal pod
+autoscaling](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) -
+that runs periodically to check if scaling the control-plane machines is
+required.
+
+In addition the implementation has to ensure that scaling decision are not based
+on single point-in-time values to prevent scaling up or down based on usage
+spikes.
+
+Instead a running average (or similar) should be used to decide about scaling.
+
+If for any reason metrics can not be gathered, the operator must remain
+available and only the autoscaling should be disabled / non-functional.
+
+### Additional RBAC permissions
+
+The operator will require access to the following resources to gather the data:
+
+- Read access for `/apis/metrics.k8s.io/v1beta1/nodes`.
+- Read access for `/api/v1/nodes` (this seems already in place for OpenShift,
+  but has to be verified for a non-OpenShift installations).
+
+### Open Questions
```

Comment:
Are there any additional metrics other than those available from metrics API that a user may want to use to scale up?

Reply:
Unfortunately not - the metrics api return exactly these two metrics based on my tests:
```
{
  "kind": "NodeMetricsList",
  "apiVersion": "metrics.k8s.io/v1beta1",
  "metadata": {},
  "items": [
    {
      "metadata": {
        "name": "minikube",
        "creationTimestamp": "2023-10-24T07:07:58Z",
        "labels": {
          "beta.kubernetes.io/arch": "amd64",
          "beta.kubernetes.io/os": "linux",
          "kubernetes.io/arch": "amd64",
          "kubernetes.io/hostname": "minikube",
          "kubernetes.io/os": "linux",
          "minikube.k8s.io/commit": "fd7ecd9c4599bef9f04c0986c4a0187f98a4396e",
          "minikube.k8s.io/name": "minikube",
          "minikube.k8s.io/primary": "true",
          "minikube.k8s.io/updated_at": "2023_09_01T12_34_46_0700",
          "minikube.k8s.io/version": "v1.31.2",
          "node-role.kubernetes.io/control-plane": "",
          "node.kubernetes.io/exclude-from-external-load-balancers": ""
        }
      },
      "timestamp": "2023-10-24T07:06:56Z",
      "window": "55.659s",
      "usage": {
        "cpu": "239401711n",
        "memory": "1421684Ki"
      }
    }
  ]
}
```

What might be worth considering is, taking into account our control plane sizing guidelines for OpenShift based on number of workers in the cluster: https://docs.openshift.com/container-platform/4.13/scalability_and_performance/recommended-performance-scale-practices/recommended-control-plane-practices.html